include/beast/core/impl/handler_ptr.ipp

```diff
@@ -0,0 +1,138 @@
+//
+// Copyright (c) 2013-2016 Vinnie Falco (vinnie dot falco at gmail dot com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+
+#ifndef BEAST_IMPL_HANDLER_PTR_HPP
+#define BEAST_IMPL_HANDLER_PTR_HPP
+
+#include <beast/core/handler_helpers.hpp>
+#include <boost/asio/detail/handler_alloc_helpers.hpp>
+#include <boost/assert.hpp>
+#include <memory>
+
+namespace beast {
+
+template<class T, class Handler>
+template<class DeducedHandler, class... Args>
+inline
+handler_ptr<T, Handler>::P::
+P(DeducedHandler&& h, Args&&... args)
+    : n(1)
+    , handler(std::forward<DeducedHandler>(h))
+{
+    t = reinterpret_cast<T*>(
+        beast_asio_helpers::
+            allocate(sizeof(T), handler));
+    try
+    {
+        t = new(t) T{handler,
+            std::forward<Args>(args)...};
+    }
+    catch(...)
+    {
+        beast_asio_helpers::
+            deallocate(t, sizeof(T), handler);
+        throw;
+    }
+}
+
+template<class T, class Handler>
+template<class DeducedHandler, class... Args>
+handler_ptr<T, Handler>::
+handler_ptr(int, DeducedHandler&& handler, Args&&... args)
+    : p_(new P(std::forward<DeducedHandler>(handler),
+        std::forward<Args>(args)...))
+{
+}
+
+template<class T, class Handler>
+handler_ptr<T, Handler>::
+~handler_ptr()
+{
+    if(! p_)
+        return;
+    if(--p_->n)
```

Comment:
Shouldn't that this be an `std::atomic`? Are we always guaranteed that this will only be invoked in a strand?